lib/ansible/galaxy/role.py

```diff
@@ -400,4 +400,14 @@ def requirements(self):
 
                     break
 
+            if type(self._requirements) is dict:
+                display.error("Wrong format for requirements for role %s" % self.name)
```

Comment:
Regarding my comment about making it an error, I was mostly wondering if we wanted to support the v2 format at all (i.e. maybe it should be fatal, like raising an AnsibleError). If collections are not supported dependencies, and we don't plan on making it so (we don't), then what is the point to supporting a more complicated format?

If we are adding support for self._requirements as a dict, then I wouldn't add the error message here, and only warn if the 'collections' key is found.

Reply:
I feel not qualified to decide this TBH.

I think a warning but continuing would be a good think for now with an option to make it fatal later?
Maybe there could just be a dedicated function / assertion for the requirements?
That would enable an upgrade at a later time
But on the other hand roles are dead, long live collections is what I understand

Reply:
@s-hertel What would you think about using the `_parse_requirements_file` from `lib/ansible/cli/galaxy.yml` ?

Reply:
@JensHeinrich I'm sorry for not responding, I lost track of this one until it came up in our backlog meeting last week. No pressure if working on this is totally off your radar - I can open another PR, but just wanted to answer here. I've spent quite a bit more time looking at this, and besides (still) thinking it should probably be an error, it is incomplete because:

* Malformed dependencies in meta/main.yml have the same issue and aren't addressed
* The requirements property doesn't consistently return the same info if it's called multiple times

> I think a warning but continuing would be a good think for now with an option to make it fatal later?

My main concern was that we shouldn't add support for new formats as a bugfix, if it's never been supported and is currently an error and adds no functional benefit. Here's the behavior I see without your patch when installing a collection with malformed meta/requirements.yml:
```
the full traceback was:

Traceback (most recent call last):
  File "/home/liveuser/ansible/bin/ansible-galaxy", line 135, in <module>
    exit_code = cli.run()
  File "/home/liveuser/ansible/lib/ansible/cli/galaxy.py", line 549, in run
    context.CLIARGS['func']()
  File "/home/liveuser/ansible/lib/ansible/cli/galaxy.py", line 75, in method_wrapper
    return wrapped_method(*args, **kwargs)
  File "/home/liveuser/ansible/lib/ansible/cli/galaxy.py", line 1170, in execute_install
    self._execute_install_role(role_requirements)
  File "/home/liveuser/ansible/lib/ansible/cli/galaxy.py", line 1257, in _execute_install_role
    role_dependencies = (role.metadata.get('dependencies') or []) + role.requirements
TypeError: can only concatenate list (not "dict") to list
```
Basically the same thing happens when installing malformed dependencies in the meta/main.yml:
```
...
TypeError: unsupported operand type(s) for +: 'dict' and 'list'
```
Unhandled tracebacks are always a bug, but raising an AnsibleError/AnsibleParserError with a helpful error would fix that.

At playbook runtime, if the dependencies in the meta/main.yml are malformed I get the (correctly handled and helpful) AnsibleParserError `ERROR! Expected role dependencies to be a list.`, so I think we should give a similar error at install time for both meta/main.yml and meta/requirements.yml for consistency.

I can't think of a good reason to support v2 format in the role's requirements.yml since it's used strictly for installing role dependencies, which should only ever be other roles. If there's a reason to expand the supported syntax, that could be a feature to add intentionally later. Adding support for it and later adding a toggle to disable the behavior again seems backwards.

> Maybe there could just be a dedicated function / assertion for the requirements?
> That would enable an upgrade at a later time

I'm not exactly sure what you mean by this - do you mean determining if there's an upgrade for a role? That's unrelated to v1 vs v2 syntax - the requirements syntax isn't the issue preventing upgrading roles. The issue is that roles don't have a mandatory versioning specification so there's no way to determine if there's a newer version. I'd review a PR that tries to implement such a thing (roles are not dead), but it would need to be backwards-compatible, mirror collection upgrade behavior, and would require replumbing role dependency resolution.

> What would you think about using the _parse_requirements_file from lib/ansible/cli/galaxy.yml and moving it to a util?

The GalaxyRole objects are created via _parse_requirements_file, so that might cause recursion issues. Unless there's a functional reason to support v2 syntax, I don't think the refactoring is worth it.

Reply:
I opened https://github.com/ansible/ansible/pull/77821 to just handle the error better.